decimal/decimal.go

```diff
@@ -0,0 +1,109 @@
+// Package decimal with support of Tarantool's decimal data type.
+//
+// Decimal data type supported in Tarantool since 2.2.
+//
+// Since: 1.6
+//
+// See also:
+//
+// * Tarantool MessagePack extensions https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/#the-decimal-type
+//
+// * Tarantool data model https://www.tarantool.io/en/doc/latest/book/box/data_model/
+//
+// * Tarantool issue for support decimal type https://github.com/tarantool/tarantool/issues/692
+//
+// * Tarantool module decimal https://www.tarantool.io/en/doc/latest/reference/reference_lua/decimal/
+package decimal
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/shopspring/decimal"
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// Decimal numbers have 38 digits of precision, that is, the total
+// number of digits before and after the decimal point can be 38.
+// A decimal operation will fail if overflow happens (when a number is
+// greater than 10^38 - 1 or less than -10^38 - 1).
+//
+// See also:
+//
+// * Tarantool module decimal https://www.tarantool.io/en/doc/latest/reference/reference_lua/decimal/
+const DecimalPrecision = 38
+
+// Decimal external type.
+const decimalExtID = 1
+
+func encodeDecimal(e *msgpack.Encoder, v reflect.Value) error {
+	number := v.Interface().(decimal.Decimal)
+	one := decimal.NewFromInt(1)
+	maxSupportedDecimal := decimal.New(1, DecimalPrecision).Sub(one) // 10^DecimalPrecision - 1
+	minSupportedDecimal := maxSupportedDecimal.Neg().Sub(one)        // -10^DecimalPrecision - 1
+	if number.GreaterThan(maxSupportedDecimal) {
+		return fmt.Errorf("msgpack: decimal number is bigger than maximum supported number (10^%d - 1)", DecimalPrecision)
+	}
+	if number.LessThan(minSupportedDecimal) {
+		return fmt.Errorf("msgpack: decimal number is lesser than minimum supported number (-10^%d - 1)", DecimalPrecision)
+	}
+
+	strBuf := number.String()
+	bcdBuf, err := EncodeStringToBCD(strBuf)
+	if err != nil {
+		return fmt.Errorf("msgpack: can't encode string (%s) to a BCD buffer: %w", strBuf, err)
+	}
+	if _, err = e.Writer().Write(bcdBuf); err != nil {
+		return fmt.Errorf("msgpack: can't write bytes to encoder writer: %w", err)
+	}
+
+	return nil
+}
+
+// Decimal values can be encoded to fixext MessagePack, where buffer
+// has a fixed length encoded by first byte, and ext MessagePack, where
+// buffer length is not fixed and encoded by a number in a separate
+// field:
+//
+// +--------+-------------------+------------+===============+
+// | MP_EXT | length (optional) | MP_DECIMAL | PackedDecimal |
+// +--------+-------------------+------------+===============+
+//
+// Before reading a buffer with encoded decimal number (PackedDecimal) we need
+// to allocate it, but before reading we don't know it's exact size.
+// msgpack.Decoder in msgpack v2 package pass a buffer with PackedDecimal bytes
+// and there is no possibility to read length in advance. to obtain MessagePack
```

Comment:
Seems like there are some trails of previous doc iteration.

Reply:
Fixed, thanks.